File: lib/api/datasets.ts
import { apiFetch } from "./client"
import type { Dataset } from "../types"

interface DatasetParams {
  category?: string
  search?: string
  sort_by?: string
  sort_dir?: string
}

export async function getDatasets(params?: DatasetParams): Promise<Dataset[]> {
  const clean: Record<string, string> = {}
  if (params) {
    for (const [key, value] of Object.entries(params)) {
      if (value) clean[key] = value
    }
  }
  const query = Object.keys(clean).length ? new URLSearchParams(clean).toString() : ""
  const path = query ? `/api/datasets?${query}` : "/api/datasets"
  const data = await apiFetch<Record<string, unknown>[]>(path)
  return data.map(mapDataset)
}

export async function getDataset(id: string): Promise<Dataset> {
  const data = await apiFetch<Record<string, unknown>>(`/api/datasets/${id}`)
  return mapDataset(data)
}

function mapDataset(d: Record<string, unknown>): Dataset {
  return {
    id: d.id as string,
    title: d.title as string,
    program: d.program as string,
    organization: d.organization as string,
    category: d.category as string,
    records: d.records as number,
    format: d.format as string,
    license: d.license as string,
    lastUpdated: d.last_updated as string,
    qualityScore: d.quality_score as number,
    downloads: d.downloads as number,
    description: d.description as string,
    tags: d.tags as string[],
  }
}

File: backend/routes/datasets.py
from typing import Optional

from fastapi import APIRouter, HTTPException

from backend.models import Dataset
from backend.data.datasets import datasets

router = APIRouter(prefix="/api/datasets", tags=["datasets"])


@router.get("", response_model=list[Dataset])
def list_datasets(
    category: Optional[str] = None,
    search: Optional[str] = None,
    sort_by: Optional[str] = None,
    sort_dir: Optional[str] = "desc",
):
    result = list(datasets)
    if category:
        result = [d for d in result if d["category"] == category]
    if search:
        q = search.lower()
        result = [
            d
            for d in result
            if q in d["title"].lower()
            or q in d["organization"].lower()
            or q in d["description"].lower()
        ]
    if sort_by and sort_by in ("last_updated", "quality_score", "downloads", "records"):
        reverse = sort_dir != "asc"
        result.sort(key=lambda d: d[sort_by], reverse=reverse)
    return result


@router.get("/{dataset_id}", response_model=Dataset)
def get_dataset(dataset_id: str):
    for d in datasets:
        if d["id"] == dataset_id:
            return d
    raise HTTPException(status_code=404, detail="Dataset not found")
